Parallelize tests

tadeboro opened the pull request:
Changes in this commit heavily refactor the test suite layout in order to simplify it and make it parallelizable. Details about the changes are included in the individual commit messages.

mancabizjak commented on HACKING.rst:
```diff
@@ -34,28 +34,40 @@ version by running::
+
+TODO: Describe the process of adding a new module.
+
    (venv) $ ansible-test sanity --python 3.7
```
With the latest changes, this can be reduced to `$ make sanity`. 

EDIT: From the name of the section, I see that this content is subject to further polishing in the near future, but I think it makes sense to modify this here and now, since we now have more powerful `Makefile`s.

mancabizjak commented on HACKING.rst:
```diff
@@ -34,28 +34,40 @@ version by running::
+TODO: Describe the process of adding a new module.
+
    (venv) $ ansible-test sanity --python 3.7
    (venv) $ ansible-test units --python 3.7
```
Similarly, can be reduced to `$ make units`

mancabizjak commented on HACKING.rst:
```diff
@@ -34,28 +34,40 @@ version by running::
+   (venv) $ cd tests/integration
+   (venv) $ molecule --base-config base.yml test -s module_asset
 
 .. note::
```
If we update L59 and L60 so that they use `make` with new targets, this warning is not needed anymore as the Makefile takes care of determining python version for us.

tadeboro replied:
Would it be OK with you if we keep this in for now? I promise I will move this into a more appropriate document under the `docs` directory as soon as possible.

Why would I like to keep this? While I really like the wrappers, I also feel that it is still immensely valuable to know how things work under the hood. Once things start to go south, knowing a thing or two about the internal workings can be a lifesaver. And yes, I know I can always dig this out of the git's history, I would instead move this content in a single commit.

mancabizjak replied:
Sure, I understand the concern and can't say I disagree, so :+1: . I brought this up more for the sake of consistency with the first part of the updated `HACKING.rst` document which now includes instructions with `make`. 

Still, once we follow up with the dedicated document under `docs`, let's not forget to at least mention the possibility of using convenience wrappers as an alternative to running commands stated here :wink:

tadeboro replied:
What I had in mind is to keep convenience wrappers in a quickstart and other casual documents that deal with tests on a higher level. Low level commands are destined for a document about trouble shooting the testing process.

mancabizjak commented on tests/integration/Makefile:
```diff
@@ -0,0 +1,17 @@
+.PHONY: roles
+roles: $(roles)
+
+.PHONY: $(modules) $(roles)
```
I wasn't aware this is possible (and now that I'm looking at it - that it's actually pretty simple to achieve) - awesome! I always found myself temporarily hard-coding the module name somewhere...

mancabizjak commented on HACKING.rst:
```diff
@@ -34,28 +34,40 @@ version by running::
-   (venv) $ cd tests/integration/modules
-   (venv) $ molecule --base-config molecule/shared/base.yml test --all
+   (venv) $ cd tests/integration
+   (venv) $ molecule --base-config base.yml test -s module_asset
```
Alternatively: `$ make molecule/module_asset`

mancabizjak commented on Makefile:
```diff
@@ -17,12 +17,38 @@ endif
 # Make sure ansible can find our collection without having to install it.
 export ANSIBLE_COLLECTIONS_PATHS := $(realpath $(CURDIR)/../../..)
 
+python_version := $(shell \
```
:+1:

aljazkosir commented on HACKING.rst:
```diff
@@ -113,3 +113,32 @@ Once we have our playbook ready, we can test our scenario by running::
+pipeline is composed of two stages: fast one and slow one.
+
+The fast stage is composed of sanity and integration tests. Once those tests
+are done executing and passing, we start execution the slow stage that runs
```
```suggestion
are done executing and passing, we start executing the slow stage that runs
```

aljazkosir commented on HACKING.rst:
```diff
@@ -113,3 +113,32 @@ Once we have our playbook ready, we can test our scenario by running::
+When we execute the tests on the CI servers, we use those measurements to
+split the tests into partitions using the *longest processing time first*
+algorithm. Implementation is stored in the `tests/integration/partition.py``
+program. Nota that this program needs to be completely deterministic or some
```
```suggestion
program. Note that this program needs to be completely deterministic or some
```

tadeboro commented:
I think I addressed all of the comments in this PR.

I also introduced a last-minute change into our Molecule configuration. Because `flake8` and `ansible-test sanity` are both checking stylistic aspects of the code in an incompatible way, I deactivated the `flake8`. I decided to deactivate it because it was only checking the files Molecule knows about while sanity check covers all of the files in our repository.

mancabizjak commented:
Thank you, looks great :+1: !
